runtime-rs: shim part

zmlcc opened the pull request:
This PR is the first one in a series of patches of rust runtime. It covers the implementation of containerd runtime v2 api, but no functionalities about containers. Releated to #3250 

`src` includes four crates:
- shim: v2 API and ttrpc services
- shim-proto: generate necessary protocol files
- virtcontainers: only a placeholder for compilation
- agent: only a placeholder for compilation

There  may be some flaws caused by refacting, and even some wrong behaviours by inproper design . Please feel free to point them out. We can fix them asap.

> Although the main patch is extracted from Alibaba Cloud's internal repository *runD*, it is still quite different from our internal implementation.  Any services provided by Alibaba Cloud **ARE NOT** based on  the open source code in this PR

fidencio commented:
@zmlcc, I appreciate that we have this code out as soon as possible and that we can start looking at it / reviewing it.
I'm adding a "do-not-merge" label as the review will take some time and we don't want this accidentally merged.

Thanks a lot for the PR! \o/

jodh-intel commented on src/runtime-rs/src/shim/src/config.rs:
```diff
@@ -0,0 +1,89 @@
+const SANDBOX_CONFIG_PATH_KEY: &str = "io.katacontainers.config_path";
+
+const DEFAULT_CONFIG_PATHS: &[&str] = &[
+    "/etc/kata-containers/configuration.toml",
+    "/usr/share/defaults/kata-containers/configuration.toml",
```
Strictly, these are defined in `./src/runtime/Makefile` and this might break in certain packaging scenarios. I think we should either get the `Makefile` to generate a tiny rust stub file containing these values as `const` variables, or use a `build.rs` to determine the values. But since these values are being used by the existing runtime, maybe this should be handled by a higher-level directory and consumed by both the go and rust runtimes?

/cc @devimc, @jcvenegas, @fidencio for thoughts.

zmlcc replied:
Leave it to later revisions due to need to generate rust stub file.  I think we may try not to modify `src/runtime` when building `src/runtime-rs`. What about using a seperate Makefile at first ?

jodh-intel commented on src/runtime-rs/src/shim/src/config.rs:
```diff
@@ -0,0 +1,89 @@
+use std::path::{Path, PathBuf};
+use virtcontainers::config::TomlConfig;
+
+const SANDBOX_CONFIG_PATH_KEY: &str = "io.katacontainers.config_path";
```
Isn't this an annotation name? If so, the variable name should reflect that.

zmlcc replied:
Fixed in the first revision

zmlcc replied:
@dgibson It follows existing Kata logic. See https://github.com/kata-containers/kata-containers/blob/39cf2b27c1cbd48cfd2f0057d04581e8e32ce9c4/src/runtime/pkg/containerd-shim-v2/create.go#L194-L204

jodh-intel commented on src/runtime-rs/src/shim/src/config.rs:
```diff
@@ -0,0 +1,89 @@
+}
+
+fn get_config_path_from_env() -> Option<String> {
+    std::env::var("KATA_CONF_FILE").ok()
```
`KATA_CONF_FILE` should be a global `const`. It might be better to pass the variable name as a parameter to this function. That makes the test code cleaner since it doesn't need to set `KATA_CONF_FILE` (which might upset the environment the tests are running in).

zmlcc replied:
Change `KATA_CONF_FILE` to a global const in the first revision. But still directly use `KATA_CONF_FILE` in `get_config_path_from_env`. I think the env name also is part of the function logic

jodh-intel commented on src/runtime-rs/src/shim/src/delete.rs:
```diff
@@ -0,0 +1,90 @@
+        match self.do_cleanup() {
+            Ok(rsp) => {
+                rsp.write_to_writer(&mut std::io::stdout())
+                    .expect("failed to write stdout");
```
Please can we remove all unecessary `expect()` and `unwrap()` calls that are not listed in the table here:

- https://github.com/kata-containers/kata-containers/blob/main/docs/code-pr-advice.md#unsafe-code

That might require changing interfaces slightly to return a `Result`.

zmlcc replied:
Removed uncessary `unwrap()`s in the first revision. But I keep using `expect` for `Mutex.lock()`. On one hand it is most unlikely to return error when acquiring a lock, on the other hand if getting error when acquiring a lock, it would means something panicked on other threads and not recoverable. Then maybe the only correct way is to panic too.

jodh-intel commented on src/runtime-rs/src/shim/src/event_publisher.rs:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    fn type_url(&self) -> String {
+        "containerd.events.TaskOOM".to_string()
```
What is this string exactly? Is it going to be valid even if the container manager is not containerd?

zmlcc replied:
Fixed in the first revision

jodh-intel commented on src/runtime-rs/src/shim/src/event_publisher.rs:
```diff
@@ -0,0 +1,220 @@
+        };
+
+        if let Err(err) = self.send_event(&e) {
+            error!(
+                "failed to send oom event for {} error {:?}",
+                container_id, err
```
Ideally, all the new code here would use structured logging. Something like the following (untested):

```rust
error!("failed to send oom event"; "container-id" => container_id, "error" => format!("{:?}", err));
```


See: https://github.com/kata-containers/tests/tree/main/cmd/log-parser

zmlcc replied:
Leave it to the later revisions. May change log libs to https://github.com/kata-containers/kata-containers/tree/main/src/libs/logging

zmlcc replied:
Fixed in the second revision: now use https://github.com/kata-containers/kata-containers/tree/main/src/libs/logging as the log libs to generate structured logging

jodh-intel commented on src/runtime-rs/src/shim/src/lib.rs:
```diff
@@ -0,0 +1,311 @@
+
+const SOCKET_ROOT: &str = "/run/containerd";
+
+const ENV_SOCKET_FD: &str = "BIND_FD";
```
Is this a standardised name? Could it be changed to `KATA_BIND_FD`?

zmlcc replied:
Fixed in the first revision

jodh-intel commented on src/runtime-rs/src/shim/src/lib.rs:
```diff
@@ -0,0 +1,311 @@
+            return Err(Error::InvalidArgument);
+        }
+        if is_delete && self.bundle.is_empty() {
+            return Err(Error::InvalidArgument);
+        }
+
+        if Self::is_component_dangerous(&self.id) || Self::is_component_dangerous(&self.namespace) {
+            return Err(Error::InvalidArgument);
```
Is there any way to add more detail to the returned error to make it clear which condition failed?

zmlcc replied:
I am not sure about this. Adding more detail would lead to the code too complicated

jodh-intel commented on src/runtime-rs/src/shim/src/lib.rs:
```diff
@@ -0,0 +1,311 @@
+        Ok(())
+    }
+
+    fn is_component_dangerous(comp: &str) -> bool {
```
I like this sort of check, but what about:

- If `comp` is `""`?
- If the path contains unicode values (including the various unicode whitespace values)?
- if the overall path container `../` or `../../../` *etc*?

I'd consider inverting the logic to only accepting printable characters and non-whitespace characters, but by using the standard facilities like https://doc.rust-lang.org/std/primitive.char.html#method.is_whitespace and we're bound to miss lots of scenarios.

zmlcc replied:
Fixed in the first revision. Use more strict rules: only allow ascii alphanumeric character and '-' and '_'

zmlcc replied:
Fixed in the second revision: now allow ascii alphanumeric character and '-', '_', '.' and '~'

jodh-intel commented on src/runtime-rs/src/shim/src/lib.rs:
```diff
@@ -0,0 +1,311 @@
+        let file_path = bundle_path.join("shim.pid");
+        // Limit the size of content to read in.
+        let size = file_path.metadata().unwrap().len();
+        if size > 32 {
```
Why?

zmlcc replied:
Here we just check the file length to "weakly" make sure the pid file does include invalid content. Change 32 to 20, which is the maximum number of decimal digits for u64

jodh-intel commented on src/runtime-rs/src/shim/src/lib.rs:
```diff
@@ -0,0 +1,311 @@
+    fn read_pid_file(&self, bundle_path: &Path) -> Result<String> {
+        let file_path = bundle_path.join("shim.pid");
+        // Limit the size of content to read in.
+        let size = file_path.metadata().unwrap().len();
```
Please can we avoid `unwrap()`.

zmlcc replied:
Fixed in the first revision

jodh-intel commented on src/runtime-rs/src/shim/src/lib.rs:
```diff
@@ -0,0 +1,311 @@
+        std::env::current_dir().map_err(Error::BundlePath)
+    }
+
+    fn socket_address(&self, id: &str) -> String {
```
- Shouldn't this return a `Path` or `PathBuf` as it's constructing a path?
- Please check to make sure `! id.is_empty()`.

zmlcc replied:
Fixed in the first revision

jodh-intel commented on src/runtime-rs/src/shim/src/lib.rs:
```diff
@@ -0,0 +1,311 @@
+    fn socket_address(&self, id: &str) -> String {
+        format!(
+            "{}/{}/{}/{}",
+            SOCKET_ROOT, &self.args.namespace, id, "shim.sock"
```
`shim.sock` should be a global `const` to make it more obvious what's being hard-coded, and to help the unit tests.

zmlcc replied:
Fixed in the first revision

jodh-intel commented on src/runtime-rs/src/shim/src/run.rs:
```diff
@@ -0,0 +1,180 @@
+impl ShimExecutor {
+    pub fn run(&mut self) {
+        if let Err(e) = self.args.validate(false) {
+            eprintln!("run shim err: {}", e);
```
Is this needed or is it just debug that can be deleted? If we need it, Is there no logger we can use at this stage? But why can't the function return a `Result` instead?

zmlcc replied:
This `run` method is used as the most outside function to handle "normal run" condition,  so it does not return errors but directly write errors from inner function calls to stderr.  The `start` and `delete` method follows the same logic

jodh-intel commented on src/runtime-rs/src/virtcontainers/src/lib.rs:
```diff
@@ -0,0 +1,22 @@
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("not found")]
```
Is this for a file? If so, is `{}` missing from the message (for the filename)?

Or should this be "not implemented"?

zmlcc replied:
Fixed in the first revision

zmlcc commented:
Thanks @fidencio .  Any feedback is welcome

jodh-intel commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,848 @@
+
+impl Task for TaskService {
+    fn state(&self, _ctx: &TtrpcContext, req: shim::StateRequest) -> Result<shim::StateResponse> {
+        debug!("====> state: {:?}", req);
+        let rsp = self.do_state(req);
+        debug!("<==== state: {:?}", rsp);
```
I'm not sure we need all these debug calls. It also relates to a missing feature: tracing, which would show all the function  calls and exits automatically if you decorated these methods with `#[instrument]`.

zmlcc replied:
Leave it to the later revisions. Using tracing is a better choice.

zmlcc replied:
I have tried `#[instrument]` , but it only create a span that include input arguments. I think it is more for tracing, but not  automatically recording input and output values. So I keep these debug calls unchanged to record more information in debug mode.

jodh-intel commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,848 @@
+        _ctx: &TtrpcContext,
+        req: shim::CreateTaskRequest,
+    ) -> Result<shim::CreateTaskResponse> {
+        debug!("====> create: {:?}", req);
+        let rsp = self.do_create(req);
+        debug!("<==== create: {:?}", rsp);
+        rsp
```
tbh, most of these functions are so similar, I think it might make sense to create a macro to generate them. I keep meaning to auto-generate more of https://github.com/kata-containers/kata-containers/blob/main/tools/agent-ctl/src/client.rs in a similar way.

zmlcc replied:
I am not sure about auto generation.  Although it would reduce similar codes,  using a macro also make it more difficult for to modify or debug a specific function.

jodh-intel commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,848 @@
+            let msg = ServerMessage::new(ServerAction::ShutdownForce);
+            self.shutdown_sender
+                .lock()
+                .expect("poisoned lock")
```
Please return a value rather than calling `expect()`.

zmlcc replied:
May using `expect()` for `Mutex.lock()`  be permitted ?

jodh-intel commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,848 @@
+                .lock()
+                .expect("poisoned lock")
+                .send(msg)
+                .unwrap();
```
Please return a value rather than calling `unwrap()`.

zmlcc replied:
Fixed `unwrap` in the first revision

jodh-intel commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,848 @@
+                .send(msg)
+                .unwrap();
+        } else {
+            debug!("<==== shutdown: cannot shutdown shim, containers not empty");
```
Is there no way to force a shutdown? The code make it look like there should be since the normal case is to specify `ServerAction::ShutdownForce`.

zmlcc replied:
This behavior is consistent with the go version kata implementation: do nothing if any container exists. See https://github.com/kata-containers/kata-containers/blob/70274b9d39e8f0ca3f8dcda710510ef8943ab2c0/src/runtime/pkg/containerd-shim-v2/service.go#L936-L943

`ServerAction::ShutdownForce` means to stop listen socket and exit proecss

jodh-intel commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,848 @@
+                    destination: String::from(mount.get_target()),
+                    options: mount.get_options().to_vec(),
+                    device_id: String::new(),
+                    // TODO:
```
This looks like it could get "lost" unless we add a GitHub issue URL here (and a unit test to check it).

zmlcc replied:
Removed unused TODO and corresponding code in the first revision.  Add them back in future release if necessary

jodh-intel commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,848 @@
+
+        sandbox
+            .signal_process(cid, eid, signal, all)
+            .map_err(|e| {
+                warn!("signal cid {} eid {} process err {:?}", cid, eid, e);
+            })
```
This looks slightly odd because if `do_exec()` fails, we return an error, but if `do_kill()` fails, we return success. 
Shouldn't the caller decide whether to ignore the error or not? Returning success for most scenarios also makes unit testing almost impossible (unless you're prepared to redirect the logger and parse the log calls).

Finally, I'm assuming the signal is included in `e`?

zmlcc replied:
Fixed in the first revision.

jodh-intel commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,848 @@
+                    "total_inactive_anon" => p_m.set_total_inactive_anon(v),
+                    "total_active_file" => p_m.set_total_active_file(v),
+                    "total_unevictable" => p_m.set_total_unevictable(v),
+                    _ => (),
```
How about logging this so we know which metrics we're not recording?

zmlcc replied:
Fixed in the first revision

jodh-intel reviewed (commented):
Thanks @zmlcc !

A few comments from a first-pass review...

mxpv commented:
Hey. Sorry for being a bit late for the party.
We (containerd) just [open sourced rust extensions](https://github.com/containerd/project/issues/82) for containerd that include lots of similar code as in this PR (grpc/ttrpc clients, protos, and shim library).
Not to derail this PR, but would you be interested to consolidate efforts?
Would you be interested to upstream missing gaps and pull it in as a crate?

jiangliu commented:
> Hey. Sorry for being a bit late for the party. We (containerd) just [open sourced rust extensions](https://github.com/containerd/project/issues/82) for containerd that include lots of similar code as in this PR (grpc/ttrpc clients, protos, and shim library). Not to derail this PR, but would you be interested to consolidate efforts? Would you be interested to upstream missing gaps and pull it in here as a crate?

Sorry, we missed your proposal (https://github.com/containerd/project/issues/82) too. But I'm really glad and excited to see that Containerd community start to provide rust extensions.

Absolutely we should consolidate our works, and hosting it by Containerd will not only help Kata, but may also help other Containerd shim-v2 implementations.

liubin commented on src/runtime-rs/rust-toolchain.toml:
```diff
@@ -0,0 +1,2 @@
+[toolchain]
+channel = "1.57.0"
```
It's better add a new line.
![image](https://user-images.githubusercontent.com/1212008/145750008-4b5c5969-55a1-455a-a311-863354361b9f.png)

liubin replied:
And about the Rust version, now we are using `versions.yaml` to manage it, if you this file, we will face keeping two places to manage. cc @jodh-intel @fidencio @bergwolf @jiangliu

zmlcc replied:
Add new line in first revision. I suggest keep rust-toolchian.toml.  Version defined in `versions.yaml` is only illustrative but in `rust-toolchian.toml` is compulsory. It would help maintain the consistency of the compilation environment

bergwolf commented:
Agreed with @mxpv that we should reuse the rust extensions to better work with containerd community. In the meantime, let's move forward with the PR and add more changes incrementally. /cc @zmlcc

zmlcc commented on src/runtime-rs/src/agent/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+thiserror = "1.0"
```
Fixed in the first revision. But I think using a full semver version may be too strict for some rust libs and may cause multiple versions for some crates in Cargo.lock

zmlcc commented on src/runtime-rs/src/shim/src/bin/containerd-shim-kata-v2.rs:
```diff
@@ -0,0 +1,101 @@
+            } else if rest_args[0] == "delete" {
+                Action::Delete(shim_args)
+            } else {
+                Action::Error(format!("unknown parameters: {}", rest_args[0]))
```
Fixed in the first revision

zmlcc commented on src/runtime-rs/src/shim/src/bin/containerd-shim-kata-v2.rs:
```diff
@@ -0,0 +1,101 @@
+        Ok(rest_args) => {
+            if help {
+                Action::Help
+            } else if rest_args.is_empty() {
+                Action::Run(shim_args)
+            } else if rest_args[0] == "start" {
+                Action::Start(shim_args)
+            } else if rest_args[0] == "delete" {
+                Action::Delete(shim_args)
```
I tried but using `match`  seems more "busy".  so keep as it is

zmlcc commented on src/runtime-rs/src/shim/src/bin/containerd-shim-kata-v2.rs:
```diff
@@ -0,0 +1,101 @@
+    -address string
+          grpc address back to main containerd
+    -bundle string
+          path to the bundle if not workdir
```
This is consistant with containerd. See https://github.com/kata-containers/kata-containers/blob/3865a1bcf696e050be4ca76c4b8c615154973d5e/src/runtime/vendor/github.com/containerd/containerd/runtime/v2/shim/shim.go#L115-L117

jodh-intel replied:
In that case, please change the message to "containerd workdir".

zmlcc replied:
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim/src/bin/containerd-shim-kata-v2.rs:
```diff
@@ -0,0 +1,101 @@
+    -bundle string
+          path to the bundle if not workdir
+    -debug
+          enable debug output in logs
```
This is consistant with containerd. See https://github.com/kata-containers/kata-containers/blob/3865a1bcf696e050be4ca76c4b8c615154973d5e/src/runtime/vendor/github.com/containerd/containerd/runtime/v2/shim/shim.go#L110-L111

zmlcc commented on src/runtime-rs/src/shim/src/bin/containerd-shim-kata-v2.rs:
```diff
@@ -0,0 +1,101 @@
+}
+
+fn main() {
+    let args = std::env::args_os().collect::<Vec<_>>();
```
As the version number is generated from `Makefile`, Leave this to later revisions until the genenration method for rust stub file is clear after discussion

zmlcc replied:
Fixed in the second revision: add `--version`

zmlcc commented:
@jodh-intel Thanks for the detailed review!  I make the first revision and please check them patiently.

I want to discuss about `expect()` and  `unwrap()`. I agree that it is imperative to minimise the use of `expect()` and  `unwrap()`. But using `expect()` for `Mutex.lock()` should be permitted. On one hand it is most unlikely to return error when acquiring a lock, on the other hand if getting error, it would means something panicked on other threads and not recoverable. Then maybe the only correct way is to panic too.

liubin commented:
/test

jodh-intel commented:
> But using expect() for Mutex.lock() should be permitted.

@zmlcc - I tend to agree. Could you consider adding that to the table in https://github.com/kata-containers/kata-containers/blob/main/docs/code-pr-advice.md#rust-language-specifics ?

jodh-intel commented:
@mxpv, @jiangliu, @bergwolf - This containerd effort sounds really interesting, and of course it makes sense to share code where possible. However, I'd like to understand how this might impact Kata 3.x and crio.

zmlcc commented:
@fidencio  I noticed there are documents reference check in CI. But in our rust code there are seperate crates (such as [shim-proto](https://github.com/zmlcc/kata-containers/tree/runtime-rs-shim/src/runtime-rs/src/shim-proto) ) that have `README.md` . These files would not be referened by any other documents. Can we modify the relevant CI check?

fidencio commented:
> @fidencio I noticed there are documents reference check in CI. But in our rust code there are seperate crates (such as [shim-proto](https://github.com/zmlcc/kata-containers/tree/runtime-rs-shim/src/runtime-rs/src/shim-proto) ) that have `README.md` . These files would not be referened by any other documents. Can we modify the relevant CI check?

Totally!

dgibson reviewed (commented):
So far this review is only for the first patch of 4.  I hope to get to the rest, but it might not be for a while.

zmlcc commented on src/runtime-rs/src/shim/protocols/src/empty.rs:
```diff
@@ -0,0 +1,157 @@
+// This file is generated by rust-protobuf 2.25.2. Do not edit
+// @generated
+
```
I think saving generated files could reduce building time. These files are not generated by default unless the built script is turned on  by  feature `generate`. See  
https://github.com/zmlcc/kata-containers/blob/ed1292fbba9aa205634c2ba02619c0e2e7a75399/src/runtime-rs/src/shim-proto/build.rs#L10

Meanwhile I notice current go version Kata also keeps a lot of generetad proto go files. Like  https://github.com/kata-containers/kata-containers/blob/1a95b6e2e0c4f29503451ffad11890a33dff84b4/src/runtime/virtcontainers/pkg/agent/protocols/grpc/oci.pb.go#L1-L3 
or
https://github.com/kata-containers/kata-containers/blob/1a95b6e2e0c4f29503451ffad11890a33dff84b4/src/runtime/vendor/github.com/gogo/protobuf/types/empty.pb.go#L1-L3

zmlcc commented on src/runtime-rs/src/shim/src/bin/main.rs:
```diff
@@ -0,0 +1,85 @@
+        }
+
+        Err(e) => Action::Error(format!("{}", e)),
+    }
```
However `clap` cannot support single hyphen in long arguments like `-id XXXX -namespace XXXX`, which is used by `containerd` from go-lang world. See https://github.com/clap-rs/clap/issues/1210

jiangliu replied:
yeah, I have tried to replace it with clap but found that doesn't work.
"go-flags" seems the only candidate for rust to deal with go flags.

jiangliu commented:
> @mxpv, @jiangliu, @bergwolf - This containerd effort sounds really interesting, and of course it makes sense to share code where possible. However, I'd like to understand how this might impact Kata 3.x and crio.

According to my understanding, the rust-extensions provides rust bindings for containerd APIs. Currently there are two major APIs implemented: Containerd GRPC API and Shim v2 API, of which shim v2 API is our interests.
When considering switching to rust-extensions, there are several factors we need to think about:
1) whether it will affect other container engines such as CRI-O? 
2) what's the mechanism to extend the shim API to support new feature of Kata v3? 

For the first question, I feel shimv2 API is some sort of de fact standard, we may stick to it.
For the second question, we may need  to invent some new prototype API such as kata-shim-extension or kata-shim-v3 or shim-v3:)

Any other more factors need to be taken into account?

jiangliu commented on src/runtime-rs/src/shim/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+edition = "2018"
+
+[[bin]]
+name = "containerd-shim-kata-v2"
```
How about "containerd-shim-kata-v3"?

zmlcc replied:
if we change name, the replacement of rust version with go version will not be seamless. And may need more works in CI

jiangliu commented on src/runtime-rs/src/agent/Cargo.toml:
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "agent"
```
It will act as "agent-service" to access services exported by the kata-agent.

dgibson replied:
Right, I realized that eventually, it's basically the "agent client".  Is there some way we can make that clearer in the name?

zmlcc replied:
Change  to "agent_client" in the second revision

jiangliu commented on src/runtime-rs/Cargo.toml:
```diff
@@ -0,0 +1,4 @@
+[workspace]
+members = [
+    "src/shim",
```
How about "crates"? Then we have "src/runtime-rs/crates/shim/src"

dgibson replied:
"crates" works.  So would just putting them directly as subdires under runtime-rs.

zmlcc replied:
Change `src` to `crates` in the second revision.  Do not put sub-crates directly as subdires under runtime-rs to distinguish some tests resouces under runtime-rs

zmlcc commented on src/runtime-rs/src/shim/src/config.rs:
```diff
@@ -0,0 +1,70 @@
+pub type ConfigResult<T> = std::result::Result<T, ConfigError>;
+
+fn get_config_path_from_oci_spec() -> Option<String> {
+    let bundle_path = &get_bundle_path().ok()?;
```
This function has been changed in latest commit.

zmlcc commented on src/runtime-rs/src/virtcontainers/src/sandbox.rs:
```diff
@@ -0,0 +1,147 @@
+    }
+
+    pub fn get_agent(&self) -> Arc<dyn Agent> {
+        self.agent.clone()
```
Because `Arc<dyn Agent>`  will be moved to a seperate thread to watch oom event. In the thread, there is no information about sandbox

dgibson replied:
Ah, ok.

zmlcc commented on src/runtime-rs/src/shim/src/delete.rs:
```diff
@@ -0,0 +1,41 @@
+    let mut rsp = shim_proto::DeleteResponse::new();
+    rsp.set_exit_status(128 + libc::SIGKILL as u32);
+    let mut exited_time = protobuf::well_known_types::Timestamp::new();
+    exited_time.set_seconds(
```
This is because `shim_proto::DeleteResponse` is an auto-generated type from proto file and includes special fields members. So we do not construct it directly but use the generated functions to set relevant value.

dgibson replied:
Ok.

zmlcc commented on src/runtime-rs/src/shim/src/delete.rs:
```diff
@@ -0,0 +1,41 @@
+            .expect("SystemTime before UNIX EPOCH!")
+            .as_secs() as i64,
+    );
+    rsp.set_exited_at(exited_time);
```
Explained above

zmlcc replied:
Fixed in the second revision: constructing the whole value.

zmlcc commented on src/runtime-rs/src/shim/src/delete.rs:
```diff
@@ -0,0 +1,41 @@
+use crate::{Error, Result, ShimArgs};
+
+pub fn do_delete(args: &ShimArgs) {
+    match cleanup(args) {
```
I do not much understand your point. In this function, if `cleanup` returns ok it will write result to stdout, otherwise it will write errors to stderr

zmlcc commented on src/runtime-rs/src/shim/src/run.rs:
```diff
@@ -0,0 +1,177 @@
+    };
+}
+
+fn long_run(args: &ShimArgs) -> Result<()> {
```
It just means this function runs as long as the whole sandbox life cycle.  Any suggestion?

zmlcc commented on src/runtime-rs/src/shim/src/run.rs:
```diff
@@ -0,0 +1,177 @@
+    for r in rx.iter() {
+        match r.action {
+            ServerAction::StartListen => {
+                if let Err(e) = server.start() {
```
`ServerAction::StartListen` and `ServerAction::StopListen` are reserved for future expansion.  E.g. we can temporarily stop to listen ttrpc socket ,  do something interesting and start to listen again  without exiting the process.  They are not used yet

zmlcc commented on src/runtime-rs/src/shim/src/start.rs:
```diff
@@ -0,0 +1,126 @@
+fn new_command(shim_args: &ShimArgs, bundle_path: &Path) -> Result<std::process::Command> {
+    let self_exec = std::env::current_exe().map_err(Error::SelfExec)?;
+    let mut command = std::process::Command::new(self_exec);
+    command
```
This follows existing Kata logic. See https://github.com/kata-containers/kata-containers/blob/1a95b6e2e0c4f29503451ffad11890a33dff84b4/src/runtime/pkg/containerd-shim-v2/service.go#L186-L206

zmlcc commented on src/runtime-rs/src/virtcontainers/src/oci.rs:
```diff
@@ -0,0 +1,82 @@
+pub fn contaier_type(
+    spec: &oci_spec::runtime::Spec,
+) -> std::result::Result<ContainerType, OciSpecInfoError> {
+    let annotations = match spec.annotations() {
```
I think maybe you are right that the container type can be identified by id. However here we just follow existing Kata logic. See https://github.com/kata-containers/kata-containers/blob/1a95b6e2e0c4f29503451ffad11890a33dff84b4/src/runtime/pkg/oci/utils.go#L319-L325

dgibson replied:
Yes, I realize that, but I'm saying that the existing logic is fragile and ugly.

zmlcc commented on src/runtime-rs/src/shim/src/delete.rs:
```diff
@@ -0,0 +1,41 @@
+
+fn cleanup(args: &ShimArgs) -> Result<shim_proto::DeleteResponse> {
+    let mut rsp = shim_proto::DeleteResponse::new();
+    rsp.set_exit_status(128 + libc::SIGKILL as u32);
```
It follows exsiting Kata logic. See https://github.com/kata-containers/kata-containers/blob/39cf2b27c1cbd48cfd2f0057d04581e8e32ce9c4/src/runtime/pkg/containerd-shim-v2/service.go#L364-L369

zmlcc commented on src/runtime-rs/src/shim/src/config.rs:
```diff
@@ -0,0 +1,70 @@
+}
+
+fn get_preferred_config_path() -> Option<PathBuf> {
+    let path = get_config_path_from_oci_spec().or_else(get_config_path_from_env)?;
```
I'm afraid not. The preferred config path should be checked whether it is None at first, then use `std::fs::canonicalize()` to check whether it exists on the file system. The default config path should be checked whether one of path lists exists on the file system. I cannot find a clear way to combine the two circumstances together directly in `load_configuration()`

zmlcc commented on src/runtime-rs/src/shim/protocols/proto/google/protobuf/any.proto:
```diff
@@ -0,0 +1,158 @@
+//
+message Any {
+  // A URL/resource name that uniquely identifies the type of the serialized
+  // protocol buffer message. This string must contain at least
```
This comes for `any.proto`, which is a dependency for `shim.proto`. See https://github.com/kata-containers/kata-containers/blob/1a95b6e2e0c4f29503451ffad11890a33dff84b4/src/runtime/vendor/github.com/containerd/containerd/runtime/v2/task/shim.proto#L20-L24

zmlcc commented on src/runtime-rs/src/shim/protocols/build.rs:
```diff
@@ -0,0 +1,29 @@
+        "proto/github.com/containerd/containerd/api/events_task.proto",
+        "proto/github.com/containerd/cgroups/stats/v1/metrics.proto",
+        "proto/google/protobuf/empty.proto",
+    ];
```
We try not to directly depend on existing Kata files.

zmlcc commented on src/runtime-rs/src/shim/protocols/Cargo.toml:
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "protocols"
```
This file is outdated. A new seperate `shim-proto` crate exists in the latest commit.

dgibson commented on src/runtime-rs/src/virtcontainers/src/oci.rs:
```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2021 Alibaba Cloud
```
That's *not* the OCI spec, though (that's [here](https://github.com/opencontainers/runtime-spec)).  Plus, that Go code has plenty of stuff that *is* related to the OCI spec, even if it has some other stuff as well.  The Rust code here has essentially nothing from the OCI spec.

zmlcc replied:
I mean it helps to extract some information from oci spec. So Change the name to "spec_info" in the second revision

dgibson commented on src/runtime-rs/src/shim-proto/LICENSE:
```diff
@@ -0,0 +1,201 @@
+                                 Apache License
```
Do you need a full license file, rather than using an SPDX identifier?

zmlcc replied:
The `shim-proto` crate is designed to be published on crate.io. So it contains seperate LICENSE file

dgibson commented on src/runtime-rs/src/shim-proto/build.rs:
```diff
@@ -4,9 +4,13 @@
 // SPDX-License-Identifier: Apache-2.0
```
This would be easier review if the patches were broken on logical grounds, rather than on the actual history of development.  You've just introduced this code, and now it's being reworked.

zmlcc replied:
Fixed in the second revision: rebased the patches by logical separations

dgibson commented on src/runtime-rs/src/shim-proto/examples/shim-proto-ttrpc-client.rs:
```diff
@@ -0,0 +1,53 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
```
In the rest of Kata we're using SPDX identifiers rather than full license banners.

zmlcc replied:
Fixed in the secode revision

dgibson commented on src/runtime-rs/src/shim-proto/examples/shim-proto-ttrpc-client.rs:
```diff
@@ -0,0 +1,53 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
What *would* be useful in the banner is some indication of what this file is for.  I think it's a debugging tool?

zmlcc replied:
the `shim-proto` is designed to be published on crates.io. So these examples files are just examples for how to use  `shim-proto` crate. They are not related to kata directly

dgibson commented on src/runtime-rs/src/shim/Cargo.toml:
```diff
@@ -30,4 +30,4 @@ slog-scope = "4.4"
 serde_json = "1.0"
-toml = "0.5"
\ No newline at end of file
+toml = "0.5"
```
This change should be folded into the original commit.

zmlcc replied:
Fixed in the second revision

dgibson commented on src/runtime-rs/src/shim/Cargo.toml:
```diff
@@ -2,32 +2,41 @@
+#categories = [""]
+repository = "https://github.com/kata-containers/kata-containers.git"
+#readme = "README.md"
+license = "Apache-2.0"
```
Why not put all this information in when the file is first created.

zmlcc replied:
Fixed in the second revision: rebased the patches by logical separations

dgibson commented on src/runtime-rs/src/shim/src/bin/containerd-shim-kata-v2.rs:
```diff
@@ -15,10 +17,11 @@ enum Action {
 }
 
-fn parse_args() -> Action {
+fn parse_args(args: &[OsString]) -> Action {
```
Again, all this re-organization makes review difficult.  Please rebase to make a *logical* separation between each patch.

zmlcc replied:
Fixed in the second revision: rebased the patches by logical separations

dgibson commented on src/runtime-rs/src/shim/src/config.rs:
```diff
@@ -57,14 +52,39 @@ fn get_default_config_path() -> Option<PathBuf> {
+        let qemu = config.hypervisor.get("qemu").unwrap();
+        assert_eq!(qemu.default_vcpus, 4);
+    }
 }
```
All the changes above here look like they're probably worthwhile.  However, they have nothing to do with the `ShimExecutor` rework.  This is what I mean by the commits not being logically separated.

zmlcc replied:
Fixed in the second revision: rebased the patches by logical separations

dgibson reviewed (commented):
This next batch of comments covers only patch 2 and a little of patch 3.

bergwolf commented:
> @mxpv, @jiangliu, @bergwolf - This containerd effort sounds really interesting, and of course it makes sense to share code where possible. However, I'd like to understand how this might impact Kata 3.x and crio.

@jodh-intel FWIW, I don't think we'll share the core logic of kata 3.x with containerd, and there is no reason to. IMO the shim API part and the shim daemon part can be shared, just like what we did in the current GO version, -- the API definition and much of the daemon process part is actually in containerd. However, how we service the SHIM APIs (which is the core of Kata Containers), remains in Kata as always.

As for CRIO, I think as long as we keep the shim API compatibility, it is not affected.

jodh-intel reviewed (commented):
Thanks @zmlcc. A few more comments.

I'm going to have to wait until January to re-review this as it's just too big a PR to review in one go. As @dgibson says splitting this into smaller more logical pieces would help get this landed more quickly imho.

zmlcc commented on src/runtime-rs/src/shim/protocols/build.rs:
```diff
@@ -0,0 +1,29 @@
+            ..Default::default()
+        })
+        .run()
+        .expect("Gen codes failed.");
```
Fixed in the seconde revision

zmlcc commented on src/runtime-rs/src/shim/src/event_publisher.rs:
```diff
@@ -0,0 +1,187 @@
+pub trait Event {
+    fn r#type(&self) -> String;
+    fn type_url(&self) -> String;
+}
```
Removed these code in the second revision

zmlcc commented on src/runtime-rs/src/shim/src/event_publisher.rs:
```diff
@@ -0,0 +1,187 @@
+    r#type: String,
+    type_url: String,
+    value: Vec<u8>,
+}
```
Removed these code in the second revision

zmlcc commented on src/runtime-rs/src/shim/src/event_publisher.rs:
```diff
@@ -0,0 +1,187 @@
+    event: EventData,
+) -> Result<()> {
+    let any = Any {
+        type_url: event.type_url.clone(),
```
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim/src/event_publisher.rs:
```diff
@@ -0,0 +1,187 @@
+    namespace: &str,
+    event: EventData,
+) -> Result<()> {
+    let any = Any {
```
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim/src/run.rs:
```diff
@@ -0,0 +1,177 @@
+    {
+        let ep = event_publisher.clone();
+        let mut ep = ep.lock().unwrap();
+        ep.start(agent, event_publisher);
```
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,904 @@
+            let mut stdin = String::from("");
+            let mut stdout = String::from("");
+            let mut stderr = String::from("");
+
```
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,904 @@
+                };
+
+                stdin = match &container.common_process.stdin {
+                    Some(stdin) => stdin.clone(),
```
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim/src/task_service.rs:
```diff
@@ -0,0 +1,904 @@
+            None
+        } else {
+            Some(req.stdin.clone())
+        };
```
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim-proto/README.md:
```diff
@@ -0,0 +1,39 @@
+# shim-proto
+
+Protobuf message and ttRPC service definitions for Containerd shim v2 protocol.
```
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim/Cargo.toml:
```diff
@@ -0,0 +1,44 @@
+keywords = ["kata containers", "shimv2"]
+#categories = [""]
+repository = "https://github.com/kata-containers/kata-containers.git"
+#readme = "README.md"
```
Fixed in the second revision

zmlcc commented on src/runtime-rs/src/shim-proto/proto/github.com/containerd/containerd/api/events_task.proto:
```diff
@@ -0,0 +1,94 @@
+/*
+	Copyright The containerd Authors.
```
These files are all under `proto` directory. I think the proto files should be stable and backward compatible, so we can just update the files as need and do not follow upstream eagerly.

zmlcc commented on src/runtime-rs/src/shim-proto/src/events_task.rs:
```diff
@@ -0,0 +1,2702 @@
+// This file is generated by rust-protobuf 2.25.2. Do not edit
```
I am not sure. We can put auto-generated code into `auto-generated/` directory, but rust can not include all .rs files under a directory automatically.  The `lib.rs` or `mod.rs`  is still to be modified manually. (Surely we can use `build.rs` or macros to make modification automatically, but I think it will be a bit more complicated)

zmlcc commented:
@dgibson  @jodh-intel  Sorry for keeping some development history in the patches.  Therefore, I rebased the whole patches into logic seperations to make it easier to review. Also fixed some flaws in this revision, includes:
- add `--version` command line options
- using structed logging
- restructe event publisher
- rewrite several code that are un-Rusty 
- change `src` folder name to `crates`
- change corresponding crate name to `agent_client` to make clearer meaning
- change corresponding file name to `spec_info` to make clearer meaning
- change the main crate name to `kata`  ( Better name suggestions are welcome)
- move `shim-proto` to `kata-containers/src/libs`, as this crates is designed for independence and could be published on [crates.io](https://crates.io/)

zmlcc commented on src/runtime-rs/src/shim/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+[package]
+name = "shim"
```
I cannot find a better name :(.    Some candidates I can think of:
- `kata`
- `containerd-shim-kata-v2`
- `rund` (successor for `runc`, also ali's internal project name)

Just change the name to "kata" in the second revision. Any better suggestion is welcome.

jodh-intel commented on src/libs/shim-proto/README.md:
```diff
@@ -0,0 +1,40 @@
+The `shim-proto` crate provides [Protobuf](https://github.com/protocolbuffers/protobuf.git) message
+and [ttRPC](https://github.com/containerd/ttrpc.git) service definitions for the
+[Containerd shim v2](https://github.com/containerd/containerd/blob/main/runtime/v2/task/shim.proto) protocol.
+Details about `shim-v2` can be found [here](https://github.com/kata-containers/kata-containers/tree/main/docs/design/architecture#shim-v2-architecture)
```
This should be a relative URL I think so that it wil work if we renamed the repo, but also will continue to work when it's copied into a stable branch.

jodh-intel commented on src/libs/shim-proto/proto/github.com/containerd/containerd/api/events_task.proto:
```diff
@@ -0,0 +1,94 @@
+import weak "github.com/containerd/containerd/protobuf/plugin/fieldpath.proto";
+
+option go_package = "github.com/containerd/containerd/api/events;events";
+option (containerd.plugin.fieldpath_all) = true;
```
What does this do?

jodh-intel commented on src/libs/shim-proto/proto/github.com/containerd/containerd/api/types/mount.proto:
```diff
@@ -0,0 +1,45 @@
+	string source = 2;
+
+	// Target path in container
+	string target = 3;
```
Can you document whether this is expected to be an absolute path below the container's root path?

jodh-intel commented on src/libs/shim-proto/proto/github.com/containerd/containerd/runtime/v2/task/shim.proto:
```diff
@@ -0,0 +1,202 @@
+	string stdin = 5;
+	string stdout = 6;
+	string stderr = 7;
+	string checkpoint = 8;
+	string parent_checkpoint = 9;
```
A comment explaining the difference between these two would be useful. Also, what format are these fields?

jodh-intel commented on src/libs/shim-proto/proto/github.com/containerd/containerd/api/types/mount.proto:
```diff
@@ -0,0 +1,45 @@
+/*
```
It would be helpful if all these files were under a `vendor/` directory to make it clearer that they are "read only" / pristine sources that should not be changed.

Also, I don't see details in the commits of which containerd commit these files are from?

jodh-intel commented on src/libs/shim-proto/src/events_task.rs:
```diff
@@ -0,0 +1,2743 @@
+        self.exited_at.take().unwrap_or_else(|| ::protobuf::well_known_types::Timestamp::new())
+    }
+
+    // string id = 5;
```
?

jodh-intel commented on src/libs/shim-proto/src/empty.rs:
```diff
@@ -0,0 +1,166 @@
+// This file is generated by rust-protobuf 2.14.0. Do not edit
+// @generated
```
Like `vendor/`, it would be helpful if these files lived in an `auto-generated/` directory I think.

jodh-intel reviewed (commented):
Thanks @zmlcc! A few comments (this is another big PR which is going to take a few reviews to look at thoroughly)...

One minor comment: I don't know about others, but I find the `texture/` directory a bit confusing. How about `tests/data/` for clarity?

jcvenegas commented on src/libs/shim-proto/README.md:
```diff
@@ -0,0 +1,40 @@
+The message and service definition source files are auto-generated from the
+[shim-v2 protobuf source file](https://github.com/containerd/containerd/blob/main/runtime/v2/task/shim.proto)
+by using [ttrpc-codegen](https://github.com/containerd/ttrpc-rust/tree/master/ttrpc-codegen). So please do not
+edit those auto-generated source files. If upgrading/modifications are needed, please follow the steps:
```
idea: having this as an script src/libs/shim-proto/scripts/update_generated_code.sh

jcvenegas commented on src/libs/shim-proto/README.md:
````diff
@@ -0,0 +1,40 @@
+shim-proto = "0.1"
+```
+
+Then add `extern crate shim-proto;` to your crate root.
````
Use `use shim-proto;` ? I am still a rust newbie in some cases, but I understand that `extern crate ...` was often used previous Rust 2018 edition.

jiangliu replied:
Most projects have been converted to edition 2018, so "use crate;" is not necessary anymore.

jcvenegas commented on src/libs/shim-proto/build.rs:
```diff
@@ -0,0 +1,35 @@
+//
+
+#[cfg(not(feature = "generate"))]
+fn main() {}
```
panic!("Not supported"); ?
When this feature option will be used?

jiangliu replied:
> panic!("Not supported"); ? When this feature option will be used?

There are two possible ways to deal with auto-generated code:
1) do not publish the auto-generated code and generate them on demand.
2) publish the auto-generated code, and only regenerate the code when needed.
Here we adopt method 2,  feature "generate" is used to regenerate code. When feature "generate" is not enabled.

jodh-intel commented:
@jiangliu - Difficult call. I'd say we should try to use the upstream files provided by containerd, but we need to:

- ensure that it's clear these files are vendored code.
- have a process to sync these vendor files periodically.

Maybe this is worth discussing at the AC meeting?

jiangliu commented:
> @jiangliu - Difficult call. I'd say we should try to use the upstream files provided by containerd, but we need to:
> 
> * ensure that it's clear these files are vendored code.
> * have a process to sync these vendor files periodically.
> 
> Maybe this is worth discussing at the AC meeting?

Yes, a good topic for AC:) 
And it would be great to sync with @mxpv too about the upgrading/maintenance policy:)

BTW, these two crates have been published with some changes needed by runD:
https://crates.io/crates/containerd-shim
https://crates.io/crates/containerd-shim-protos

jodh-intel reviewed (commented):
Thanks @zmlcc. A few more comments...

zmlcc commented on src/libs/shim-proto/Cargo.toml:
```diff
@@ -0,0 +1,33 @@
+[package]
```
Thanks @jodh-intel.  However, I would switch to `containerd-shim-protos` to deal with proto files. The `src/libs/shim-proto` will be removed in the next rebase.

zmlcc commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+}
+
+impl ShimArgs {
+    /// Check the shim argument object is vaild or not.
```
Nice catch. Will fixed in the next revision.

zmlcc commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+    match time.duration_since(SystemTime::UNIX_EPOCH) {
+        Err(e) => Err(Error::SystemTime(e)),
+        Ok(n) => Ok(Timestamp {
+            seconds: n.as_secs() as i64,
+            nanos: n.subsec_nanos() as i32,
```
However, `Timestamp` is generated from `google/protobuf/timestamp.proto`, in which the seconds and nanos are defined as signed int.

jodh-intel replied:
oh well :sob:

zmlcc commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+        Ok(())
+    }
+
+    fn is_component_dangerous(comp: &str) -> bool {
```
`namespace` here stands for [containerd namespace](https://github.com/containerd/containerd/blob/main/docs/namespaces.md), not linux namespace. I will try to find out how containerd verify this string later

jodh-intel replied:
@zmlcc - ok, but for now can you use `verify_cid()` as that's been used for a while and has extensive tests.

zmlcc commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+const SHIM_SOCKET: &str = "shim.sock";
+const KATA_BIND_FD: &str = "KATA_BIND_FD";
+
+/// Received commandline arguments or environment arguments.
+///
+/// For defail information, please refer to the
+/// [shim v2 spec](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md).
+#[derive(Debug, Default, Clone)]
+pub struct ShimArgs {
```
Actually all the members in this type come from command line flags required by containerd shim v2 api. They are all arguments

jodh-intel replied:
Ah! Looking at this again, I see what this is :smile:

zmlcc commented:
@jodh-intel  @jiangliu  @mxpv The `libs/shim-proto` was removed in the latest commit. Use [https://crates.io/crates/containerd-shim-protos](https://crates.io/crates/containerd-shim-protos) instead to provide protobuf message.


Besides, as we plan to use async features on ttrpc and shim, the code published here may have major changes.

fidencio commented:
A head's up that may affect your PR.

In the Architecture Committee meeting from January 25th, 2022, the Architecture Committee has agreed on using the "Dismiss stale pull request approvals when new commits are pushed" configuration from GitHub.  It basically means that if your PR has been rebased or updated, the approvals given will be erased.

In order to minimize traumas due to the new approach, please, consider adding a note on the changes done before the rebase / force-push, and also pinging the reviewers for a subsequent round of reviews.

Thanks for your understanding!

Related issue: https://github.com/kata-containers/community/issues/249

jodh-intel commented on src/runtime-rs/crates/kata/src/config.rs:
```diff
@@ -0,0 +1,87 @@
+}
+
+fn get_config_path_from_env() -> Option<String> {
+    std::env::var("KATA_CONF_FILE").ok()
```
I'd consider making this a global `const` for visibility.

jodh-intel commented on src/runtime-rs/crates/kata/src/delete.rs:
```diff
@@ -0,0 +1,89 @@
+        let bundle_path = dir.path();
+        std::env::set_current_dir(bundle_path).unwrap();
+
+        let id: String = thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(16)
+            .map(char::from)
+            .collect();
+        let namespace: String = thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(16)
+            .map(char::from)
+            .collect();
```
Might be worth creating a test utility function here to create random IDs.

jodh-intel commented on src/runtime-rs/crates/kata/src/delete.rs:
```diff
@@ -0,0 +1,89 @@
+        let executor = ShimExecutor::new(args);
+
+        let resp = executor.do_cleanup().unwrap();
+        //assert_eq!(resp.pid > 0);
```
Commented out code.

jodh-intel commented on src/runtime-rs/crates/kata/src/event_publisher.rs:
```diff
@@ -0,0 +1,203 @@
+/// Event publisher to send events to the Containerd.
+///
+/// It will automatically forward all TaskOOM events from agent to Containerd.
+/// TODO: support grace shutdown.
```
- s/grace/graceful/ ?
- The runtime will need the ability to shut down gracefully for tracing to work, so if this difference between the 2.x runtime and the 3.x runtime needs to be captured somewhere.

jodh-intel commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+pub type Result<T> = std::result::Result<T, Error>;
+
+const SOCKET_ROOT: &str = "/run/containerd";
+const SHIM_SOCKET: &str = "shim.sock";
```
Please can you add a comment explaining what this shim socket is and who owns it?

jodh-intel commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+
+const SOCKET_ROOT: &str = "/run/containerd";
+const SHIM_SOCKET: &str = "shim.sock";
+const KATA_BIND_FD: &str = "KATA_BIND_FD";
```
Is this an environment variable? If so, I'd consider adding `_ENV_VAR` to the name.

jodh-intel commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+const SHIM_SOCKET: &str = "shim.sock";
+const KATA_BIND_FD: &str = "KATA_BIND_FD";
+
+/// Received commandline arguments or environment arguments.
```
```suggestion
/// Received commandline arguments or environment arguments from a shimv2 container manager such as containerd.
```

jodh-intel commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+            .metadata()
+            .map_err(|e| Error::QueryFile(e, file_path.clone()))?
+            .len();
+        // 20 is the maximum number of decimal digits for u64
+        if size > 20 {
```
I'd rather just try to convert the string value to a u64 and return an error if that fails. That would be clearer and safer imho. It would also assure you that what you read was actually a number and not some random junk (which currently isn't handled).

jodh-intel commented on src/runtime-rs/crates/kata/src/run.rs:
```diff
@@ -0,0 +1,186 @@
+    // So it does not return errors but directly write errors to stderr
+    pub fn run(&mut self) {
+        if let Err(e) = self.args.validate(false) {
+            eprintln!("run shim err: {}", e);
+            return;
+        }
+
+        if let Err(e) = self.long_run() {
+            eprintln!("run shim error: {}", e);
```
Inconsistent error message format (`err` / `error`). I'd also consider prefixing with `"ERROR: "` to make it "stand out" (and more consistent for `grep`-ability ;)

It _might_ be worth adding a `self.real_run()` or similar so that this method could become something like:

```rust
pub fn run(&mut self) {
    if let Err(e) = self.real_run() {
        eprintln!("ERROR: {:#}", e);
    }
}
```

That looks simpler and also means that `real_run()` can be unit tested more easily.

jodh-intel commented on src/runtime-rs/crates/kata/src/run.rs:
```diff
@@ -0,0 +1,186 @@
+impl ShimExecutor {
+    // Implement ttrpc call from containerd
+    // This functions should be the most outside one for "normal run" command action
+    // So it does not return errors but directly write errors to stderr
```
Nit: Missing period at end of sentence.

jodh-intel commented on src/runtime-rs/crates/kata/src/run.rs:
```diff
@@ -0,0 +1,186 @@
+            slog::Level::Info
+        };
+
+        let (logger, async_guard) = logging::create_logger("kata-runtime-rs", "rund", level, fifo);
```
Why `rund` and not `runk`? :smile:

jodh-intel commented on src/runtime-rs/crates/kata/src/start.rs:
```diff
@@ -0,0 +1,229 @@
+    // So it does not return errors but directly write errors to stderr
+    pub fn start(&mut self) {
+        if let Err(e) = self.args.validate(false) {
+            eprintln!("start shim err: {} with invalid inputs {:?}", e, self.args);
+            return;
+        }
+
+        match self.do_start() {
+            Ok(address) => {
+                std::io::stdout()
+                    .write_all(address.as_os_str().as_bytes())
+                    .expect("failed to write stdout");
+            }
+            Err(e) => eprintln!("start shim err: {}", e),
```
Same command as above: inconsistent error format so maybe creating a helper method that returns a `Result` would be clearer and make the code simpler here?

jodh-intel commented on src/runtime-rs/crates/kata/src/task_service.rs:
```diff
@@ -0,0 +1,874 @@
+            let (exit_code, exit_time) = exit_info(container.common_process.status.clone());
+
+            let stdin = container.common_process.stdin.clone().unwrap_or_default();
+            let stdout = container.common_process.stdout.clone().unwrap_or_default();
+            let stderr = container.common_process.stderr.clone().unwrap_or_default();
+            let terminal = container.common_process.terminal;
+            Ok(shim::StateResponse {
+                id: cid.into(),
+                bundle,
+                pid: self.pid,
+                status,
+                stdin,
+                stdout,
+                stderr,
+                terminal,
+                exit_status: exit_code,
+                exited_at: exit_time.into(),
+                ..Default::default()
+            })
+        } else if let Some(exec) = container.processes.get(eid) {
+            let status = to_status(exec.state);
+            let (exit_code, exit_time) = exit_info(exec.common_process.status.clone());
+            let stdin = exec.common_process.stdin.clone().unwrap_or_default();
+            let stdout = exec.common_process.stdout.clone().unwrap_or_default();
+            let stderr = exec.common_process.stderr.clone().unwrap_or_default();
+            let terminal = exec.common_process.terminal;
+            Ok(shim::StateResponse {
+                id: cid.into(),
+                bundle,
+                pid: self.pid,
+                status,
+                stdin,
+                stdout,
+                stderr,
+                terminal,
+                exit_status: exit_code,
+                exited_at: exit_time.into(),
+                ..Default::default()
+            })
+        } else {
```
There is a lot of commonality in these two blocks so I wonder if this could be simplified by calling a helper method?

larrydewey replied:
I agree. Pulling out the generic functionality here would make this a lot easier to follow, and would prevent some duplication.

jodh-intel commented on src/runtime-rs/crates/kata/src/task_service.rs:
```diff
@@ -0,0 +1,874 @@
+        let stderr = (!req.stderr.is_empty()).then(|| req.stderr.clone());
+        let terminal = req.get_terminal();
+
+        //FIXME
```
?

jodh-intel commented on src/runtime-rs/crates/kata/src/task_service.rs:
```diff
@@ -0,0 +1,874 @@
+        let height = req.get_height();
+        let width = req.get_width();
+
+        let mut sandbox = self.sandbox.lock().expect("poisoned sandbox lock");
```
I wonder if it's worth creating a `get_sandbox!()` macro like `get_controller_or_return_singular_none()` to do this in every function?

jodh-intel reviewed (commented):
Thanks @zmlcc. A few comments...

larrydewey commented on src/runtime-rs/crates/kata/src/task_service.rs:
```diff
@@ -0,0 +1,874 @@
+    fn do_update(&self, req: shim::UpdateTaskRequest) -> Result<Empty> {
+        let cid = req.get_id();
+
+        //FIXME
```
What are the details of the needed fix?

larrydewey commented on src/runtime-rs/crates/virtcontainers/src/spec_info.rs:
```diff
@@ -0,0 +1,82 @@
+    for k in CRI_CONTAINER_TYPE_KEY_LIST.iter() {
+        if let Some(type_value) = annotations.get(*k) {
+            match type_value.as_str() {
+                "sandbox" => return Ok(ContainerType::PodSandbox),
```
Where the majority of the return values are `Ok(ContainerType::PodSandbox)`, it would probably simplify things to set the default return value at the top in a variable, and for the unique instances change the variable; breaking, and then returning the variable instead of implementing a number of returns.

pmores commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+    ///
+    /// The id, namespace, address and publish_binary are mandatory for START, RUN and DELETE.
+    /// And bundle is mandatory for DELETE.
+    pub fn validate(&mut self, is_delete: bool) -> Result<()> {
```
@jodh-intel I _think_ `is_delete` is meant to signal to `validate()` whether or not the validation is to be performed for a delete Action, because if so, `validate()` needs also to validate the bundle arg which isn't required otherwise.

But let's wait for an authoritative reply. :-)

pmores commented on src/runtime-rs/crates/kata/src/lib.rs:
```diff
@@ -0,0 +1,318 @@
+    }
+
+    fn get_bundle_path(&self) -> Result<PathBuf> {
+        std::env::current_dir().map_err(Error::BundlePath)
```
Shouldn't this function prefer to return `self.args.bundle` and only fall back to `current_dir()` if `self.args.bundle` is empty?

zmlcc commented:
> Hi @zmlcc - any update on this?

Hi @jodh-intel, we are currently working on another version of the code including async features on ttrpc and shim, which will be published in the next few days. Then this PR  will be obsoleted by the new one. And any comments menthioned here will be fixed in the new code

jodh-intel commented:
Thanks for explaining @zmlcc!

zmlcc commented:
This PR has been deprecated by #3786.  Thanks to all reviewers @dgibson @liubin @jiangliu @jodh-intel @jcvenegas @larrydewey @pmores.  The relevant comments mentioned here have been fixed in the new PR.
